### DISCUSSION TITLE
Clock with precise seconds

### DISCUSSION THREAD

**Discussion by lilynet336** [FelixKratz/SketchyBar#454] [Q&A] [answered]
I just installed SketchyBar yesterday, oh boy it's amazing and I can already say I'm never going back. My single gripe is that a clock with seconds can't be updated in realtime (because of the need to call `date`). Is there a way to tell SketchyBar to run the `date` command exactly on the second?
Or, instead, is there a more dynamic way of getting the time (in seconds) that doesn't require calling something? Sort of like an event subscription, but more of a system-level thing?

**Comment by FelixKratz** [comment]
Have you considered setting the `update_freq` of the item to `1`, such that the command is executed every second?
```bash
sketchybar --add item time right \
  --set time update_freq=1 script='sketchybar --set $NAME label="$(date "+%H:%M:%S")"'
```

Or do you want to have actual sub-second precision with the display of the seconds, without invoking a command at all?
If this is the case and you don't mind getting your hands dirty with some C, you can create a helper program, which you set up with a high precision timer and periodically send the time update directly via mach messages to sketchybar (see for example my C helper program for a hint how to do that: https://github.com/FelixKratz/dotfiles/tree/master/.config/sketchybar/helper). Doing it this way will *drastically* reduce the overhead of spawning the shell process and interpreting the script every second, with a total latency between helper and sketchybar lower than 100 microseconds.

**Comment by lilynet336** [comment]
I did set the `update_freq` to 1, but the problem is this isn't always going to execute on the second, for example, if the config finishes reloading at 10:09:30.9, then the clock will be .9 seconds late, and yes I realize this is nitpicking, but I do frequently use the seconds on the stock Menu Bar. I certainly don't mind writing some code but I'm afraid I don't know C. I will take a look at the helper program and see what I can do.

**Comment by FelixKratz** [comment, marked as answer]
Actually, it is really simple to create such C program:
```c
#include "sketchybar.h"
#include <CoreFoundation/CoreFoundation.h>
#include <time.h>

void callback(CFRunLoopTimerRef timer, void* info) {
  time_t current_time;
  time(&current_time);
  const char* format = "%H:%M:%S";
  char buffer[64];
  strftime(buffer, sizeof(buffer), format, localtime(&current_time));

  uint32_t message_size = sizeof(buffer) + 64;
  char message[message_size];
  snprintf(message, message_size, "--set time label=\"%s\"", buffer);
  sketchybar(message);
}

int main() {
  CFRunLoopTimerRef timer = CFRunLoopTimerCreate(kCFAllocatorDefault, (int64_t)CFAbsoluteTimeGetCurrent() + 1.0, 1.0, 0, 0, callback, NULL);
  CFRunLoopAddTimer(CFRunLoopGetMain(), timer, kCFRunLoopDefaultMode);
  sketchybar("--add item time right");
  CFRunLoopRun();
  return 0;
}
```
Note that the "magic" happens in the `sketchybar.h` file, which is this one: https://github.com/FelixKratz/dotfiles/blob/master/.config/sketchybar/helpers/event_providers/sketchybar.h
but you don't need to understand what is happening in there, you only need it to compile the program with:
```bash
clang -std=c99 clock.c -framework CoreFoundation -o clock
```
and then execute it with:
```bash
./clock
```

You can of course automate this to autmatically start with sketchybar (see my dotfiles for reference).
With this running you will need no scripts at all and it will be perfectly in sync with the actual time.

**Comment by lilynet336** [comment]
Wow, thank you! I follow the logic of what you wrote, not that I'd be able to write C myself, but I see what you're saying now. For some reason though it still doesn't seem to be realtime. The timer still seems to rely on when `clock` starts running, because depending on when through the second I run it, the delay varies. I tried adding one second to the `current_time` in the code, but then it (of course) gets ahead of the normal clock by a little bit. I also tried setting the timer in the C program to execute every 0.5 seconds (possibly improperly) to no avail. I'm not really sure what I may be doing wrong here so I'm including the relevant config in case it helps. I did modify the names and paths of a few things, but that shouldn't matter (right?).

sketchybarrc:
```
sketchybar --add item clock left
sketchybar --set clock padding_right=5 background.drawing=1 icon.background.drawing=1 icon="􀐫" icon.color=0xff000000 icon.padding_left=4 icon.padding_right=5 background.border_width=1 label.padding_left=5 label.padding_right=5 background.color=0x77000000 label.font.size=12.5
# The below is the compiled clock.c
~/.config/sketchybar/scripts/clock &
```

Then in `clock.c`, I removed the `sketchybar --add` call so it could be handled in the `sketchybarrc` instead, renamed the `item` to `clock`, and changed the format of the time string.

**Comment by FelixKratz** [comment]
Ah, I forgot about that. You need to round down the `CFAbsoluteTimeGetCurrent()` result from a double precision number to an integer, such that you fire the timer exactly at the begin of a new second:
```c
  CFRunLoopTimerRef timer = CFRunLoopTimerCreate(kCFAllocatorDefault, (int64_t)CFAbsoluteTimeGetCurrent() + 1.0, 1.0, 0, 0, callback, NULL);
```
I have updated this in my original response as well.

**Comment by lilynet336** [comment]
That line did the trick! Thank you again, I really appreciate you guiding me through the solution.

**Reply by RainVier** [reply]
Hey, first of all thank you for asking this so that we have solution for that. 
Ive noticed that due to the different size of each number the clock is "moving" a little bit. Did you figure it out how to make them stable or not movable

**Comment by qns7** [comment]


**Reply by FelixKratz** [reply]
One thing to be aware of when not using the header from my answer is that you might create a lot of overhead. The header sends the message to sketchybar directly via the kernel, your version however first spawns a new process, which then sends the message to sketchybar.

Another thing is that syncing should not be neccessary because of this code in the original answer: `(int64_t)CFAbsoluteTimeGetCurrent() + 1.0`
It already makes sure that the timer is invoked on an actual second change, regardless of the time the program is started.

**Reply by qns7** [reply]
Thank you! I will look into this further this evening. :-)
EDIT: @FelixKratz  The link to your header from above is dead, which is why I scrambled my solution together, because I thought you have a totally new setup as of today. But can I still find your header file? Or which lines of your [new one](https://github.com/FelixKratz/dotfiles/blob/master/.config/sketchybar/helpers/event_providers/sketchybar.h) are the ones I would need? Because I will absolutely rock your setup then. 🎉

**Reply by FelixKratz** [reply]
I have updated the link to the header file. It is the same one. Just copy the entire header.

**Comment by vu1xan** [comment]
@lyra56k can you let me know how to implement it i can't figure out how to call clock.c and get it working

**Reply by lilynet336** [reply]
Hey,
It's been a while since I've messed with this project, but once you compile clock.c into a binary you can just stick a reference to that binary in your SketchyBar files.

**Reply by vu1xan** [reply]
i struggle with the referencing part in SketchyBar files

**Reply by lilynet336** [reply]
Sorry, had to get back to my computer. If you've already compiled `clock.c` into the binary, you can do something like this in your `sketchybarrc`:

```
sketchybar --add item clock left
sketchybar --set clock label.font.size=12.5
~/.config/sketchybar/scripts/clock &
```
The clock binary will call sketchybar itself for updates to the time.

**Reply by vu1xan** [reply]
thank you very much :)
it's working but somehow my `click_script` doesn't work with this clock

**Reply by lilynet336** [reply]
No problem :)
I'm not really sure about the click script, if the documentation isn't working maybe consider opening a new issue. I haven't used this in a long time and I don't remember much about how it works.
